### PR TITLE
Prepare release notes for 1.29+ck2 and now 1.29+ck3

### PR DESCRIPTION
**Pull request by addyess**
Preparing release-note for charmed-kubernetes 1.29+ck2

Roughly based on the bugs fixed in the [milestone](https://launchpad.net/charmed-kubernetes/+milestone/1.29+ck2)

The adjustment to the release-node generator templates reflects the current release-notes pages.

### CODE REVIEW

File: pages/k8s/how-to-cos-lite.md [hunk 1]
```diff
@@ -109,12 +109,12 @@ Deploy the grafana-agent:
 Juju deploy grafana-agent
 ```
 
-Relate `grafana-agent` to `k8s`, `kubernetes-control-plane` and `kubernetes-worker`:
+Relate `grafana-agent` to charmed kubernetes applications:
```
Comment (addyess) [inline pages/k8s/how-to-cos-lite.md:112]: replaces `k8s` with `kubeapi-load-balancer` as i'm sure the original author mistakenly wrote here.

File: pages/k8s/release-notes.md [hunk 2]
```diff
@@ -14,6 +14,31 @@ toc: False
 ---
 
 <!-- AUTOGENERATE RELEASE NOTES HERE -->
```
Comment (evilnick) [inline pages/k8s/release-notes.md:16]: This needs a blank line otherwise the web renderer messes everything up

**Review by evilnick — COMMENTED** [review]
Thanks - i couldn't make sense of where to add a line to the template to fix this forever...

**Comment by kwmonroe** [comment]
Setting to draft to make sure we include keystone / vault / ceph upgrade notes.

**Comment by addyess** [comment]
I feel bad i didn't make it clear this was THE change to for the 1.29+ck2 release.  Nice to have a chat with you both so we know what is necessary to close out last weeks' release

File: pages/k8s/upgrade-notes.md [hunk 3]
```diff
@@ -24,6 +24,201 @@ any of the intervening steps.
...
+LMA Relations:
+
+* `nrpe-external-master` (provides: `nrpe-external-master` on KCP and KW)
+* `prometheus` (provides: `prometheus-manual` on KCP)
+* `scrape` (provides: `prometheus` on KW)
+* `grafana` (provides: `grafana-dashboard` )
+
+In order to prepare for observability, see the [Integration with COS Lite
+docs][cos] which can be performed following an upgrade of the charms but prior
+to upgrade of the kubernetes cluster.
```
Comment (evilnick) [inline pages/k8s/upgrade-notes.md:50]:
```suggestion
to upgrade of the Kubernetes cluster.
```

File: pages/k8s/upgrade-notes.md [hunk 4]
```diff
@@ -24,6 +24,201 @@ any of the intervening steps.
...
+
+### kube-api-endpoint relation dropped
+
+The `kubernetes-control-plane:kube-api-endpoint` and
+`kubernetes-worker:kube-api-endpoint` relations have been removed since these
+APIs are are provided by the kube-control relation. Ensure these two apps are
```
Comment (evilnick) [inline pages/k8s/upgrade-notes.md:56]:
```suggestion
APIs are are provided by the `kube-control` relation. Ensure these two apps are
```

File: pages/k8s/upgrade-notes.md [hunk 5]
```diff
@@ -24,6 +24,201 @@ any of the intervening steps.
...
+juju integrate kubernetes-control-plane:loadbalancer-internal kubeapi-loadbalancer
+juju integrate kubernetes-control-plane:loadbalancer-external kubeapi-loadbalancer
+juju remove-relation kubernetes-control-plane:loadbalancer kubeapi-loadbalancer
+```
+
+### ceph-client relation deprecated
+
+The `kubernetes-control-plane:ceph-client` relation is being deprecated.
+
+Investment in Ceph integration continues, but in the `ceph-csi` charm
```
Comment (evilnick) [inline pages/k8s/upgrade-notes.md:79]: 'Investment' seems like an odd choice. Perhaps something like this is clearer?

```suggestion
Ceph integration is still a priority, but continues with the `ceph-csi` charm
```

File: pages/k8s/upgrade-notes.md [hunk 6]
```diff
@@ -24,6 +24,201 @@ any of the intervening steps.
...
+which integrates Ceph with Kubernetes.
+
+After upgrading the `kubernetes-control-plane` charm, the charm
+may enter `blocked` status with the message:
+`ceph-client relation deprecated, use ceph-csi charm instead`.
+
+If you see this message, you can resolve it by removing the ceph-client
```
Comment (evilnick) [inline pages/k8s/upgrade-notes.md:86]:
```suggestion
If you see this message, you can resolve it by removing the `ceph-client`
```

File: pages/k8s/upgrade-notes.md [hunk 7]
```diff
@@ -24,6 +24,201 @@ any of the intervening steps.
...
+juju integrate ceph-csi kubernetes-control-plane
+juju integrate ceph-csi ceph-mon
+juju remove-relation kubernetes-control-plane:ceph-client ceph-mon
+```
+
+### Keystone/K8s Authentication management
+
+Charmed Kubernetes was installing and managing an older version of 
+keystone-auth which manages authentication and authorisation
+through keystone.
```
Comment (evilnick) [inline pages/k8s/upgrade-notes.md:100]:
```suggestion
through Keystone.
```

File: pages/k8s/upgrade-notes.md [hunk 8]
```diff
@@ -24,6 +24,201 @@ any of the intervening steps.
...
+
+This service is better suited to be managed externally from the
+`kubernetes-control-plane` charm. However, the charm provides the following
+upgrade method to maintain the deployment of this service beyond 1.28.
+
+One can determine if Keystone management is applicable with
```
Comment (evilnick) [inline pages/k8s/upgrade-notes.md:106]:
```suggestion
One can determine if Keystone management is applicable with:
```

File: pages/k8s/upgrade-notes.md [hunk 9]
```diff
@@ -24,6 +24,201 @@ any of the intervening steps.
...
+
+```
+keystone:identity-credentials  kubernetes-control-plane:keystone-credentials   keystone-credentials    regular
+```
+
+...then you'll need to prepare a bit before the upgrade.
+
+#### Resources
+
+One should familiarise themself with k8s-keystone auth via the [upstream docs][keystone-auth]
```
Comment (evilnick) [inline pages/k8s/upgrade-notes.md:124]: ?
```suggestion
The [upstream Keystone docs][keystone-auth] cover keystone-auth in detail and should be the main reference for implementation details.
```

File: pages/k8s/upgrade-notes.md [hunk 10]
```diff
@@ -24,6 +24,201 @@ any of the intervening steps.
...
+should be enabled. This command adds the Keystone authorisation webhook config
+and the `Webhook` authorisation mode:
+
+``` 
+juju config kubernetes-control-plane \
+    authorization-webhook-config-file="$(cat keystone-upgrade/webhook.yaml)" \
+    authorization-mode="Node,RBAC,Webhook"
+```
+
+Finally, acknowledge the charm no longer manages keystone by removing the relation:
```
Comment (evilnick) [inline pages/k8s/upgrade-notes.md:170]:
```suggestion
Finally, acknowledge the charm no longer manages Keystone by removing the relation:
```

**Comment by evilnick** [comment]
I'm not done with this but just making some running tweaks. How close are we to having all the bits?

**Comment by kwmonroe** [comment]
> I'm not done with this but just making some running tweaks. How close are we to having all the bits?

@evilnick i think we're good2go here. I did notice the `release-notes-headers.j2` template removed some quotes and modified the yaml structure, but the `1.29/release-notes.md` puts them back.  I'm not sure which way is supposed to be the source of truth there, but I trust you'll if an update is needed to either :)

**Comment by evilnick** [comment]
I'm calling this done for now, I may pick up some other things.
I don't really like how complicated this is and that we require people to look at a load of notes, but I don't see a clean way to do it. Thanks for all your work on this.

**Pull request merged**